File: src/spellchecker/SpellChecker.cxx
/* Openoffice.org-voikko: Finnish linguistic extension for OpenOffice.org
 * Copyright (C) 2007 Harri Pitkänen <hatapitk@iki.fi>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *********************************************************************************/

#include <hspell.h>

#include "SpellChecker.hxx"
#include "SpellAlternatives.hxx"
#include "../common.hxx"

namespace hspell {

SpellChecker::SpellChecker(uno::Reference<uno::XComponentContext> const & context) :
	cppu::WeakComponentImplHelper5
	     <lang::XServiceInfo,
	      linguistic2::XSpellChecker,
	      linguistic2::XLinguServiceEventBroadcaster,
	      lang::XInitialization,
	      lang::XServiceDisplayName>(m_aMutex),
	compContext(context) { }

OUString SAL_CALL SpellChecker::getImplementationName() throw (uno::RuntimeException) {
	return getImplementationName_static();
}

sal_Bool SAL_CALL SpellChecker::supportsService(const OUString & serviceName)
	throw (uno::RuntimeException) {
	uno::Sequence<OUString> serviceNames = getSupportedServiceNames();
	for (sal_Int32 i = 0; i < serviceNames.getLength(); i++)
		if (serviceNames[i] == serviceName) return sal_True;
	return sal_False;
}

uno::Sequence<OUString> SAL_CALL SpellChecker::getSupportedServiceNames() throw (uno::RuntimeException) {
	return getSupportedServiceNames_static();
}

uno::Sequence<lang::Locale> SAL_CALL SpellChecker::getLocales() throw (uno::RuntimeException) {
	uno::Sequence<lang::Locale> locales(1);
	locales.getArray()[0] = lang::Locale(A2OU("he"), A2OU("IL"), OUString());
	return locales;
}

sal_Bool SAL_CALL SpellChecker::hasLocale(const lang::Locale & aLocale) throw (uno::RuntimeException) {
	if (aLocale.Language == A2OU("he")) return sal_True;
	else return sal_False;
}

sal_Bool SAL_CALL SpellChecker::isValid(const OUString & aWord, const lang::Locale &,
	                              const uno::Sequence<beans::PropertyValue> & aProperties)
	throw (uno::RuntimeException, lang::IllegalArgumentException) {
	osl::MutexGuard vmg(getVoikkoMutex());
	if (!voikko_initialized) return sal_False;
	OString oWord = OUStringToOString(aWord, RTL_TEXTENCODING_ISO_8859_8);
	const char * c_str = oWord.getStr();

	thePropertyManager->setValues(aProperties);
	// VOIKKO_DEBUG_2("SpellChecker::isValid: c_str: '%s'\n", c_str);
	int preflen;
	int result = hspell_check_word(dict, c_str, &preflen);
	// VOIKKO_DEBUG_2("SpellChecker::isValid: result = %i\n", result);
	thePropertyManager->resetValues(aProperties);
	if (result) return sal_True;
	else return sal_False;
}

uno::Reference<linguistic2::XSpellAlternatives> SAL_CALL SpellChecker::spell(
	const OUString & aWord, const lang::Locale &,
	const uno::Sequence<beans::PropertyValue> & aProperties)
	throw (uno::RuntimeException, lang::IllegalArgumentException) {
	
	// Check if diagnostic message should be returned
	if (aWord.equals(A2OU("VoikkoGetStatusInformation"))) {
		SpellAlternatives * alternatives = new SpellAlternatives();
		alternatives->word = aWord;
		uno::Sequence<OUString> suggSeq(1);
		if (thePropertyManager != 0)
			suggSeq.getArray()[0] = thePropertyManager->getInitializationStatus();
		else
			suggSeq.getArray()[0] = A2OU("PropertyManager does not exist");
		alternatives->alternatives = suggSeq;
		return alternatives;
	}
	
	osl::MutexGuard vmg(getVoikkoMutex());
	if (!voikko_initialized) return 0;
	OString oWord = OUStringToOString(aWord, RTL_TEXTENCODING_ISO_8859_8);
	const char * c_str = oWord.getStr();

	thePropertyManager->setValues(aProperties);
	int preflen;
	if (hspell_check_word(dict, c_str, &preflen)) {
		thePropertyManager->resetValues(aProperties);
		return 0;
	}

	struct corlist cl;
	corlist_init(&cl);
	hspell_trycorrect(dict, c_str, &cl);
	thePropertyManager->resetValues(aProperties);
	SpellAlternatives * alternatives = new SpellAlternatives();
	alternatives->word = aWord;
	int scount = corlist_n(&cl);
	if (scount == 0) return alternatives;
	uno::Sequence<OUString> suggSeq(scount);
	OUString * suggStrings = suggSeq.getArray();
	OString ostr;
	for (int i = 0; i < scount; i++) {
		ostr = OString(corlist_str(&cl, i));
		suggStrings[i] = OStringToOUString(ostr, RTL_TEXTENCODING_ISO_8859_8);
	}
	corlist_free(&cl);

	alternatives->alternatives = suggSeq;
	return alternatives;
}

sal_Bool SAL_CALL SpellChecker::addLinguServiceEventListener(
	const uno::Reference<linguistic2::XLinguServiceEventListener> & xLstnr)
	throw (uno::RuntimeException) {
	osl::MutexGuard vmg(getVoikkoMutex());
	VOIKKO_DEBUG("SpellChecker::addLinguServiceEventListener");
	if (thePropertyManager != 0)
		return thePropertyManager->addLinguServiceEventListener(xLstnr);
	else return sal_False;
}

sal_Bool SAL_CALL SpellChecker::removeLinguServiceEventListener(
	const uno::Reference<linguistic2::XLinguServiceEventListener> & xLstnr)
	throw (uno::RuntimeException) {
	osl::MutexGuard vmg(getVoikkoMutex());
	VOIKKO_DEBUG("SpellChecker::removeLinguServiceEventListener");
	if (thePropertyManager != 0)
		return thePropertyManager->removeLinguServiceEventListener(xLstnr);
	else return sal_False;
}

void SAL_CALL SpellChecker::initialize(const uno::Sequence<uno::Any> &)
	throw (uno::RuntimeException, uno::Exception) {
	osl::MutexGuard vmg(getVoikkoMutex());
	VOIKKO_DEBUG("SpellChecker::initialize");
	if (thePropertyManager == 0) thePropertyManager = new PropertyManager(compContext);
	thePropertyManager->initialize();
}

OUString SAL_CALL SpellChecker::getServiceDisplayName(const lang::Locale & aLocale)
	throw (uno::RuntimeException) {
	return A2OU("Hebrew spellchecker (Hspell)");
}

void SAL_CALL SpellChecker::disposing() {
	VOIKKO_DEBUG("SpellChecker::disposing");
}

}

File: oxt/pythonpath/LODivvun/SpellChecker.py
# Libreoffice-divvun: Linguistic extension for LibreOffice
# Copyright (C) 2015 Harri Pitkänen <hatapitk@iki.fi>
#
# This Source Code Form is subject to the terms of the Mozilla Public License,
# v. 2.0. If a copy of the MPL was not distributed with this file, You can
# obtain one at http://mozilla.org/MPL/2.0/.
# 
# Alternatively, the contents of this file may be used under the terms of
# the GNU General Public License Version 3 or later (the "GPL"), in which
# case the provisions of the GPL are applicable instead of those above.

import logging
import unohelper 		# type:ignore
from com.sun.star.linguistic2 import XSpellChecker, XLinguServiceEventBroadcaster  # type:ignore
from com.sun.star.lang import XServiceInfo, XInitialization, XServiceDisplayName  # type:ignore
from LODivvun.DivvunHandlePool import DivvunHandlePool
from LODivvun.SpellAlternatives import SpellAlternatives
from LODivvun.PropertyManager import PropertyManager

class SpellChecker(unohelper.Base, XServiceInfo, XSpellChecker, XLinguServiceEventBroadcaster, XInitialization, XServiceDisplayName):

	def __init__(self, ctx, *args):
		logging.debug("SpellChecker.__init__")

	# From XServiceInfo
	def getImplementationName(self):
		return SpellChecker.IMPLEMENTATION_NAME

	def supportsService(self, serviceName):
		return serviceName in self.getSupportedServiceNames()

	def getSupportedServiceNames(self):
		return SpellChecker.SUPPORTED_SERVICE_NAMES

	# From XSupportedLocales
	def getLocales(self):
		return DivvunHandlePool.getInstance().getSupportedSpellingLocales()

	def hasLocale(self, aLocale):
		return DivvunHandlePool.getInstance().supportsSpellingLocale(aLocale)

	# From XSpellChecker
	def isValid(self, word, locale, properties):
		DivvunHandlePool.mutex.acquire()
		try:
			divvun = DivvunHandlePool.getInstance().getHandle(locale)
			if divvun is None:
				return False
			PropertyManager.getInstance().setValues(properties)
			result = divvun.spell(word)
			PropertyManager.getInstance().resetValues(properties)
			return result
		finally:
			DivvunHandlePool.mutex.release()

	def spell(self, word, locale, properties):
		# Check if diagnostic message should be returned
		if word == "DivvunGetStatusInformation":
			suggestions = [DivvunHandlePool.getInstance().getInitializationStatus()]
			return SpellAlternatives(word, suggestions, locale)
		
		DivvunHandlePool.mutex.acquire()
		try:
			divvun = DivvunHandlePool.getInstance().getHandle(locale)
			if divvun is None:
				return None

			PropertyManager.getInstance().setValues(properties)
			if divvun.spell(word):
				PropertyManager.getInstance().resetValues(properties)
				return None
			suggestions = divvun.suggest(word)
			PropertyManager.getInstance().resetValues(properties)
			return SpellAlternatives(word, suggestions, locale)
		finally:
			DivvunHandlePool.mutex.release()

	# From XLinguServiceEventBroadcaster
	def addLinguServiceEventListener(self, xLstnr):
		logging.debug("SpellChecker.addLinguServiceEventListener")
		DivvunHandlePool.mutex.acquire()
		try:
			return PropertyManager.getInstance().addLinguServiceEventListener(xLstnr)
		finally:
			DivvunHandlePool.mutex.release()


	def removeLinguServiceEventListener(self, xLstnr):
		logging.debug("SpellChecker.removeLinguServiceEventListener")
		DivvunHandlePool.mutex.acquire()
		try:
			return PropertyManager.getInstance().removeLinguServiceEventListener(xLstnr)
		finally:
			DivvunHandlePool.mutex.release()

	# From XInitialization
	def initialize(self, seq):
		pass

	# From XServiceDisplayName
	def getServiceDisplayName(self, locale):
		if locale.Language == "fi":
			return "Oikoluku (Divvun)"
		else:
			return "Spellchecker (Divvun)"

SpellChecker.IMPLEMENTATION_NAME = "divvun.SpellChecker"
SpellChecker.SUPPORTED_SERVICE_NAMES = ("com.sun.star.linguistic2.SpellChecker",)
